snapshot/gc/manifest.go

```diff
@@ -0,0 +1,108 @@
+package gc
+
+import (
+	"context"
+	"encoding/json"
+	"sort"
+	"time"
+
+	"github.com/kopia/kopia/repo"
+	"github.com/kopia/kopia/repo/content"
+	"github.com/kopia/kopia/repo/manifest"
+)
+
+// ContentPrefix is the prefix used for GC related content
+const ContentPrefix content.ID = "g"
+
+const markManifestType = "gc-mark"
+
+// MarkManifest represents information about a single GC mark run
+type MarkManifest struct {
+	ID manifest.ID `json:"-"`
+
+	StartTime time.Time `json:"startTime"`
+	EndTime   time.Time `json:"endTime"`
+
+	DetailsID content.ID `json:"detailsID"`
+}
+
+// MarkDetails contains metadata about a GC mark phase, which is later used
+// by the GC delete phase
+type MarkDetails struct {
+	// Deleted snapshots encountered by a GC run. This is not strictly needed
+	// in the GC delete phase. Currently used for debugging, may be removed.
+	DeletedSnapshots []manifest.ID `json:"deletedSnapshots,omitempty"`
+	// Snapshots that were visible to a GC run.
+	Snapshots []manifest.ID `json:"liveSnapshots,omitempty"`
+	// Set of contents marked for deletion by the GC mark phase.
+	MarkedContent []content.ID `json:"markedContent,omitempty"`
+}
+
+func markContentsDeleted(ctx context.Context, rep *repo.Repository, snaps []manifest.ID, toDelete []content.ID) error {
+	if err := markContentAndCreateManifest(ctx, rep, snaps, toDelete); err != nil {
+		return err
+	}
+
+	return rep.Content.Flush(ctx)
+}
+
+func markContentAndCreateManifest(ctx context.Context, rep *repo.Repository, snaps []manifest.ID, toDelete []content.ID) error {
+	// create mark details manifest, get back an id
+	m := MarkManifest{
+		StartTime: time.Now().UTC(),
+	}
+
+	did, err := writeMarkDetails(ctx, rep, snaps, toDelete)
```

Comment:
Yes: `writeMarkDetails` is an "intent" log and it is also written asynchronously.

There are multiple invariants / guarantees provided by the store that make this work:
* `MarkDetails` is written to a single store "content". Contents are (a) immutable and (b) atomic, that is they are guaranteed to be stored in a blob and not spread across blobs, so either it is completely written or not.
* The content store guarantees that the actual content is made persistent before the index (metadata) used to find it. This means that if the writing of the mark manifest below succeeds (really in the call to to `rep.Content.Flush()`), then the mark details are guaranteed to have been written as well.
* The disabling and re-enabling of index flushing below is to ensure that all the content deletions along with the index entry for the GC mark manifest end up in a single index pack. This ensures that a concurrent snapshotting processes either sees all deletions or not at all . Otherwise, we could end up in a situation where some of the deletes are visible to a snapshot process, while the rest are not visible, and the concurrent snapshot could end up reusing it. The current implementation does not rely on this invariant, but this is needed for a later optimization in the GC repair and purge phase to prune the set of snapshots that need to be repaired.
* The motivation for flushing batches is to avoid accumulating too much state in memory, and also make (some of the) deletions visible as early as possible.

As mentioned offline, I'm in the process of cleaning up the GC notes and putting them together in a form that helps explain the overall design, assumptions and so on.